File: core/src/main/java/me/liam/queue/ActionQueue.java
package me.liam.queue;

import android.os.Handler;
import android.os.Looper;

import java.util.LinkedList;
import java.util.Queue;

public class ActionQueue {

    private Queue<Action> actionQueue = new LinkedList<>();

    private Handler handler;

    public ActionQueue(Handler handler){
        this.handler = handler;
    }

    public void enqueue(final Action action){
        if (action.actionType() == Action.TYPE_LOAD){
            if (Thread.currentThread() == Looper.getMainLooper().getThread()){
                action.run();
            }else {
                handler.post(new Runnable() {
                    @Override
                    public void run() {
                        action.run();
                    }
                });
            }
            return;
        }
        if (!actionQueue.isEmpty()) return;
        actionQueue.add(action);
        handlerAction();
    }


    private void handlerAction(){
        if (actionQueue.size() == 1){
            handler.post(new Runnable() {
                @Override
                public void run() {
                    Action a = actionQueue.peek();
                    if (a != null){
                        long duration = a.run();
                        handler.postDelayed(new Runnable() {
                            @Override
                            public void run() {
                                actionQueue.clear();
                            }
                        },duration);
                    }else {
                        actionQueue.clear();
                    }
                }
            });
        }
    }
}

File: core/src/main/java/me/liam/fragmentation/queue/ActionQueue.kt
package me.liam.fragmentation.queue

import android.os.Handler
import android.os.Looper

import java.util.LinkedList

class ActionQueue(private val handler: Handler) {

    private val actionQueue = LinkedList<Action>()

    fun enqueue(action: Action) {
        if (action.actionType() == Action.TYPE_LOAD) {
            if (Thread.currentThread() === Looper.getMainLooper().thread) {
                action.run()
            } else {
                handler.post { action.run() }
            }
            return
        }
        if (!actionQueue.isEmpty()) return
        actionQueue.add(action)
        handlerAction()
    }


    private fun handlerAction() {
        if (actionQueue.size == 1) {
            handler.post {
                val a = actionQueue.peek()
                if (a != null) {
                    val duration = a.run()
                    handler.postDelayed({ actionQueue.clear() }, duration)
                } else {
                    actionQueue.clear()
                }
            }
        }
    }
}
